// @aws-cdk/aws-batch/test/integ.batch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.app = void 0;
const ec2 = require("@aws-cdk/aws-ec2");
const ecr = require("@aws-cdk/aws-ecr");
const ecs = require("@aws-cdk/aws-ecs");
const cdk = require("@aws-cdk/core");
const batch = require("../lib/");
exports.app = new cdk.App();
const stack = new cdk.Stack(exports.app, 'batch-stack');
const vpc = new ec2.Vpc(stack, 'vpc');
const launchTemplate = new ec2.CfnLaunchTemplate(stack, 'ec2-launch-template', {
    launchTemplateName: 'EC2LaunchTemplate',
    launchTemplateData: {
        blockDeviceMappings: [
            {
                deviceName: '/dev/xvdcz',
                ebs: {
                    encrypted: true,
                    volumeSize: 100,
                    volumeType: 'gp2',
                },
            },
        ],
    },
});
new batch.JobQueue(stack, 'batch-job-queue', {
    computeEnvironments: [
        {
            computeEnvironment: new batch.ComputeEnvironment(stack, 'batch-unmanaged-compute-env', {
                managed: false,
            }),
            order: 1,
        },
        {
            computeEnvironment: new batch.ComputeEnvironment(stack, 'batch-demand-compute-env-launch-template', {
                managed: true,
                computeResources: {
                    type: batch.ComputeResourceType.ON_DEMAND,
                    vpc,
                    launchTemplate: {
                        launchTemplateName: launchTemplate.launchTemplateName,
                    },
                    computeResourcesTags: {
                        'compute-env-tag': '123XYZ',
                    },
                },
            }),
            order: 2,
        },
        {
            computeEnvironment: new batch.ComputeEnvironment(stack, 'batch-spot-compute-env', {
                managed: true,
                computeResources: {
                    type: batch.ComputeResourceType.SPOT,
                    vpc,
                    bidPercentage: 80,
                },
            }),
            order: 3,
        },
    ],
});
const repo = new ecr.Repository(stack, 'batch-job-repo');
new batch.JobDefinition(stack, 'batch-job-def-from-ecr', {
    container: {
        image: new ecs.EcrImage(repo, 'latest'),
    },
});
new batch.JobDefinition(stack, 'batch-job-def-from-', {
    container: {
        image: ecs.ContainerImage.fromRegistry('docker/whalesay'),
    },
});
